File: Word Search II/mysolution_tle.cpp
class Solution {
public:
    vector<vector<int>> dirs{{1,0},{-1,0},{0,1},{0,-1}};
    
    bool backTrack(vector<vector<char>>& board, int row, int col, string word, int index)
    {
        if(index==word.size())
            return true;
        
        int n=board.size(), m=board[0].size();
        char c=word[index];
        
        if(row<0||row>=n||col<0||col>=m||board[row][col]!=c)
            return false;
        
        board[row][col]='#';
        
        for(auto dir:dirs)
        {
            int x=row+dir[0], y=col+dir[1];
            if(backTrack(board,x,y,word,index+1))
            {
                board[row][col]=c;
                return true;
            }
        }
        
        board[row][col]=c;
        return false;
    }
    
    bool findWord(vector<vector<char>>& board, string word)
    {
        int n=board.size(), m=board[0].size();
        
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(backTrack(board,i,j,word,0))
                    return true;
            }
        }
        
        return false;
    }
    
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        
        vector<string> ans;
        
        for(auto s:words)
        {
            if(findWord(board,s))
                ans.push_back(s);
        }
        
        return ans;
    }
};

File: Word Search II/mysolution_TLE.py
class Solution:
    
    def __init__(self):
        self.n=0
        self.m=0
        self.visited=None
        self.dirs=[[1,0],[-1,0],[0,1],[0,-1]]
    
    def backTrack(self,board,word,index,row,col):
        if index==len(word):
            return True
        
        c=word[index]
        
        if row<0 or row>=self.n or col<0 or col>=self.m or self.visited[row][col] or board[row][col]!=c:
            return False
        
        self.visited[row][col]=True
        for dir1 in self.dirs:
            x=row+dir1[0]
            y=col+dir1[1]
            if self.backTrack(board,word,index+1,x,y):
                self.visited[row][col]=False
                return True
        
        self.visited[row][col]=False
        return False
            
    
    def find(self,board,word):
        for i in range(self.n):
            for j in range(self.m):
                if self.backTrack(board,word,0,i,j):
                    return True
        return False
    
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        self.n=len(board)
        self.m=len(board[0])
        self.visited=[[False for j in range(self.m)] for i in range(self.n)]
        
        ans=[]
        
        for word in words:
            if self.find(board,word):
                ans.append(word)
        
        return ans